lib/query_builder/extension.ex

```diff
@@ -0,0 +1,101 @@
+defmodule QueryBuilder.Extension do
+  @moduledoc ~S"""
+  Use this module to create an extension module to `QueryBuilder` for app specific query utilities.
+  Use your query builder extension module wherever you would normally use `QueryBuilder`
+
+  Example:
+  ```
+  defmodule MyApp.QueryBuilder do
+    use QueryBuilder.Extension
+
+    defmacro __using__(opts) do
```

Comment:
We require the user to add that __using__ but actually we can inject that too, right?

Reply:
The user technically doesn't need to make this a macro and `use MyApp.QueryBuilder`.  They can continue with `use QueryBuilder` in their models, and in the query functions utilize `MyApp.QueryBuilder.my_custom_function()`.  Also my macro does more than just use the `QueryBuilder`.  A condensed version of my extension module:

```elixir
defmodule MyApp.QueryBuilder do
  use QueryBuilder.Extension
  require Ecto.Query
  import MyApp.QueryFunctions

  defmacro __using__(opts) do
    register_full_text_search_options(opts)

    quote do
      require QueryBuilder
      QueryBuilder.__using__(unquote(opts))
    end
  end

  # Add app specific query functions
  # ---------------------------------
  def search(query, search_term), do: do_search(query, search_term)
end
```

So I personally feel this shouldn't be as prescriptive as to what the user's macro functionality should be

Reply:
@mathieuprog ^^^